// @nexbus/common/dist/services/base-service.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.BaseService = void 0;
const tslib_1 = require("tslib");
const axios_1 = tslib_1.__importDefault(require("axios"));
const Utils = tslib_1.__importStar(require("../utils/axios"));
class BaseService {
    constructor(options) {
        this.options = options;
        this.client = axios_1.default.create({
            baseURL: options.baseURL,
            headers: options.headers
        });
    }
    async get(path, config = { params: {}, headers: {}, responseType: 'json' }) {
        return Utils.get(this.client, path, config);
    }
    async post(path, data = {}, config = { params: {}, headers: {} }) {
        return Utils.post(this.client, path, data, config);
    }
    async put(path, data = {}, config = { params: {}, headers: {} }) {
        return Utils.put(this.client, path, data, config);
    }
    async patch(path, data = {}, config = { params: {}, headers: {} }) {
        return Utils.patch(this.client, path, data, config);
    }
    async del(path) {
        return Utils.del(this.client, path);
    }
}
exports.BaseService = BaseService;
